### PR TITLE
refactor: add params to multisig exported methods

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: actors/builtin/multisig/multisig_actor.go [hunk 1]
```diff
@@ -60,114 +62,188 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
...
 	st.NextTxnID += 1
+
+	txn := MultiSigTransaction{
+		Proposer:   callerAddr,
+		Expiration: 0, // TODO lotus does not specify this.
+		To:         params.To,
+		Method:     params.Method,
+		Params:     params.Params,
+		Value:      0, // TODO lotus does not specify this.
+	}
```
Comment (frrist) [inline actors/builtin/multisig/multisig_actor.go]: https://github.com/filecoin-project/lotus/blob/https://github.com/filecoin-project/lotus/commit/8418464d91d8eb7e5a2765668073579291912dbb/chain/actors/actor_multisig.go#L196

File: actors/builtin/multisig/multisig_actor.go [hunk 2]
```diff
@@ -60,116 +63,181 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
 	return h, state
 }
 
-func (a *MultiSigActor) Propose(rt vmr.Runtime, txn MultiSigTransaction) TxnID {
+type ConstructorParams struct {
+	AuthorizedParties  autil.ActorIDSetHAMT
+	NumApprovalsThreshold int64
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Include UnlockDuration here and initialize it in the constructor as Lotus does

File: actors/builtin/multisig/multisig_actor.go (hunk 1 above)
Comment (anorth) [reply actors/builtin/multisig/multisig_actor.go]: See https://github.com/filecoin-project/specs-actors/issues/5

I suggest including Expiration in the propose params for now so it's internally consistent.

Value should be set to the value or rt.ValueReceived() and *not* be an explicit parameter, sounds like a bug in Lotus (cc @whyrusleeping)

File: actors/builtin/multisig/multisig_actor.go [hunk 3]
```diff
@@ -60,116 +63,181 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
...
+type SwapAuthorizedPartyParams struct {
+	From addr.Address
+	To addr.Address
+}
+
+func (a *MultiSigActor) SwapAuthorizedParty(rt vmr.Runtime, params SwapAuthorizedPartyParams) {
 	// Can only be called by the multisig wallet itself.
 	rt.ValidateImmediateCallerIs(rt.CurrReceiver())
 
+	oldActor, err := addr.IDFromAddress(params.From)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Name these `oldParty`, `newParty`.

File: actors/builtin/multisig/multisig_actor.go [hunk 4]
```diff
@@ -57,117 +59,185 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
...
+
+func (a *MultiSigActor) Constructor(rt vmr.Runtime, params *ConstructorParams) {
+
+	rt.ValidateImmediateCallerIs(builtin.InitActorAddr)
+	h := rt.AcquireState()
+
+	st := MultiSigActorState{
+		AuthorizedParties:     params.AuthorizedParties,
+		NumApprovalsThreshold: params.NumApprovalsThreshold,
+		UnlockDuration:        params.UnlockDuration,
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Please follow the lotus code more closely:

```
	if params.UnlockDuration != 0 {
		self.InitialBalance = vmctx.Message().Value
		self.UnlockDuration = params.UnlockDuration
		self.StartingBlock = vmctx.BlockHeight()
	}
```

File: actors/builtin/multisig/multisig_actor.go [hunk 5]
```diff
@@ -57,117 +59,185 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
...
 	// https://github.com/filecoin-project/specs-actors/issues/7
 	return txnID
 }
 
-func (a *MultiSigActor) Approve(rt vmr.Runtime, txnID TxnID, txn MultiSigTransaction) {
+type TxnIDParams struct {
+	ID TxnID
+}
+
+func (a *MultiSigActor) Approve(rt vmr.Runtime, params TxnIDParams) {
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Change all params to be accepted as pointers.

File: actors/builtin/multisig/multisig_actor.go [hunk 6]
```diff
@@ -57,117 +59,189 @@ func (a *MultiSigActor) State(rt Runtime) (vmr.ActorStateHandle, MultiSigActorSt
...
 	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
 	callerAddr := rt.ImmediateCaller()
 	a._rtValidateAuthorizedPartyOrAbort(rt, callerAddr)
-	a._rtApproveTransactionOrAbort(rt, callerAddr, txnID, txn)
+	// TODO implement cancel logic
+}
+
+type ModifyAuthorizedPartyParams struct {
+	AuthorizedParty addr.Address // must be an ID protocol address.
+	ModifyPartySize bool
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: This is unused. Remove it now, follow up with a new PR to use it.

**Pull request merged**